druid/src/widget/images.rs

```
@@ -0,0 +1,259 @@
+#[derive(Clone)]
+pub struct ImageData {
+    pixels: Vec<u8>,
+    x_pixels: i32,
+    y_pixels: i32,
+}
+
+impl ImageData {
+    /// Create an empty Image
+    pub fn empty() -> Self {
+        ImageData {
+            pixels: [].to_vec(),
+            x_pixels: 0,
+            y_pixels: 0,
+        }
+    }
+
+    pub fn from_data(raw_image: Vec<u8>) -> Self {
+        let dec = image::load_from_memory(&raw_image[..]).unwrap().to_rgb();
+
+        let sizeofimage = dec.dimensions();
+        ImageData {
+            pixels: dec.to_vec(),
+            x_pixels: sizeofimage.0 as i32,
+            y_pixels: sizeofimage.1 as i32,
+        }
+    }
+
+    /// Convert ImageData into Piet draw instructions
+    pub fn to_piet(&self, scale: f64, offset: Point, paint_ctx: &mut PaintCtx) {
+        let offset_matrix = Affine::new([scale, 0., 0., scale, offset.x, offset.y]);
+
+        paint_ctx
+            .with_save(|ctx| {
+                ctx.transform(offset_matrix);
+
+                let im = ctx
+                    .make_image(
+                        self.x_pixels as usize,
+                        self.y_pixels as usize,
+                        &self.pixels,
+                        ImageFormat::Rgb,
+                    )
+                    .unwrap();
+                let rec = Rect::from_origin_size(
+                    (0.0, 0.0),
+                    (self.x_pixels as f64, self.y_pixels as f64),
+                );
+                ctx.draw_image(&im, rec, InterpolationMode::Bilinear);
+
+                Ok(())
+            })
+            .unwrap();
+    }
+}
+
+impl Default for ImageData {
+    fn default() -> Self {
+        ImageData::empty()
+    }
+}
+
+impl FromStr for ImageData {
+    type Err = Box<dyn Error>;
+
+    fn from_str(image_str: &str) -> Result<Self, Self::Err> {
+        //pub fn from_path(path: ) -> Self {
+        let image_data = image::open(image_str).unwrap().to_rgb();
```

yea, we should just return this error.